src/hepattn/models/decoder.py

```diff
@@ -284,27 +293,31 @@ def forward(
         """Forward pass for the decoder layer.
 
         Args:
-            q: Query embeddings.
-            kv: Key/value embeddings.
+            queries: Query embeddings.
+            keys: Key/value embeddings.
             attn_mask: Optional attention mask.
             q_mask: Optional query mask.
             kv_mask: Optional key/value mask.
             query_posenc: Optional query positional encoding.
             key_posenc: Optional key positional encoding.
             attn_mask_transpose: Optional transposed attention mask.
-
-        Returns:
-            Tuple of updated query and key/value embeddings.
         """
-        if query_posenc is not None:
-            q = q + query_posenc
-        if key_posenc is not None:
-            kv = kv + key_posenc
-
-        # Update query/object embeddings with the key/constituent embeddings
-        q = self.q_ca(q, kv=kv, attn_mask=attn_mask, q_mask=q_mask, kv_mask=kv_mask)
-        q = self.q_sa(q, q_mask=q_mask)
-        q = self.q_dense(q)
+        if self.skip_first_layer_pe:
+            queries = queries + self.q_sa(self.self.norm_sa(queries), kv=self.self.norm_sa(queries), v=self.self.norm_sa(queries), q_mask=q_mask)
```

Comment:
Would be good to revert changes to the order of operations here and more closely follow what we had originally, only changing the PE handling. We don't do SA first since the queries don't contain any useful info in the first layer, so it's a pointless operation